File: src/im_agent/platforms/webhook.py
from __future__ import annotations

from typing import Any

from im_agent.models import InboundMessage, OutboundMessage
from im_agent.platforms.base import PlatformAdapter


class WebhookAdapter(PlatformAdapter):
    """Generic adapter for a normalized webhook format.

    This lets us exercise the gateway immediately, even before any real IM
    platform is wired in. Later adapters can keep the same internal contract.
    """

    name = "webhook"

    def normalize_inbound(self, payload: dict[str, Any]) -> InboundMessage:
        platform = str(payload.get("platform") or "webhook").strip()
        chat_id = str(payload.get("chat_id") or "").strip()
        user_id = str(payload.get("user_id") or "anonymous").strip()
        text = str(payload.get("text") or "").strip()
        message_id = str(payload.get("message_id") or "").strip()
        chat_type = str(payload.get("chat_type") or "p2p").strip().lower() or "p2p"
        route_key = str(payload.get("route_key") or "").strip()
        session_id = str(payload.get("session_id") or "").strip()
        mentions = payload.get("mentions") or []
        attachments = payload.get("attachments") or []
        metadata = payload.get("metadata") or {}

        if not chat_id:
            raise ValueError("Payload must include chat_id")
        if not text:
            raise ValueError("Payload must include text")

        return InboundMessage(
            platform=platform,
            chat_id=chat_id,
            user_id=user_id,
            text=text,
            message_id=message_id,
            chat_type=chat_type,
            route_key=route_key,
            session_id=session_id,
            mentions=[item for item in mentions if isinstance(item, dict)],
            attachments=[item for item in attachments if isinstance(item, dict)],
            metadata=dict(metadata) if isinstance(metadata, dict) else {},
            raw_payload=payload,
        )

    def build_delivery_payload(self, outbound: OutboundMessage) -> dict[str, Any]:
        payload = super().build_delivery_payload(outbound)
        payload["delivery"] = "webhook"
        return payload

File: src/adapters/webhook.rs
use crate::adapters::PlatformAdapter;
use crate::error::GatewayError;
use crate::models::{utc_now_iso, InboundMessage, OutboundMessage};
use async_trait::async_trait;
use serde_json::{json, Value};

pub struct WebhookAdapter;

#[async_trait]
impl PlatformAdapter for WebhookAdapter {
    fn name(&self) -> &str {
        "webhook"
    }

    fn normalize_inbound(&self, payload: Value) -> Result<InboundMessage, GatewayError> {
        let object = payload
            .as_object()
            .ok_or_else(|| GatewayError::validation("Payload must be a JSON object"))?;
        let platform = str_field(object, "platform").unwrap_or_else(|| "webhook".to_string());
        let chat_id = str_field(object, "chat_id").unwrap_or_default();
        let user_id = str_field(object, "user_id").unwrap_or_else(|| "anonymous".to_string());
        let text = str_field(object, "text").unwrap_or_default();
        if chat_id.trim().is_empty() {
            return Err(GatewayError::validation("Payload must include chat_id"));
        }
        if text.trim().is_empty() {
            return Err(GatewayError::validation("Payload must include text"));
        }
        Ok(InboundMessage {
            platform,
            chat_id,
            user_id,
            text,
            message_id: str_field(object, "message_id").unwrap_or_default(),
            chat_type: str_field(object, "chat_type").unwrap_or_else(|| "p2p".to_string()),
            route_key: str_field(object, "route_key").unwrap_or_default(),
            session_id: str_field(object, "session_id").unwrap_or_default(),
            mentions: value_array(object.get("mentions")),
            attachments: value_array(object.get("attachments")),
            metadata: object
                .get("metadata")
                .and_then(Value::as_object)
                .cloned()
                .unwrap_or_default(),
            timestamp: utc_now_iso(),
            raw_payload: Value::Object(object.clone()),
        })
    }

    async fn send_outbound(&self, outbound: OutboundMessage) -> Result<Value, GatewayError> {
        Ok(json!({
            "platform": outbound.platform,
            "chat_id": outbound.chat_id,
            "text": outbound.text,
            "attachments": outbound.attachments,
            "timestamp": outbound.timestamp,
            "metadata": outbound.metadata,
            "delivery": "webhook",
            "sent": false,
        }))
    }

    fn profile(&self) -> Value {
        json!({
            "adapter_name": "webhook",
            "surface_family": "webhook",
            "transport_mode": "normalized",
            "supports_mentions": false,
            "supports_attachments": true,
            "supports_replies": true,
            "supports_updates": false,
            "supports_live_receive": false,
        })
    }
}

fn str_field(object: &serde_json::Map<String, Value>, key: &str) -> Option<String> {
    object
        .get(key)
        .and_then(Value::as_str)
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty())
}

pub(crate) fn value_array(value: Option<&Value>) -> Vec<Value> {
    value
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter(|item| item.is_object())
                .cloned()
                .collect()
        })
        .unwrap_or_default()
}
